### PR TITLE
Improve scan time performance

### PR DESCRIPTION
**Pull request by fabiosangregorio**
This PR brings three separate improvements to the scan process, resulting in an average decrease of roughly 50% in scan time (from ~53s to ~25s in my tests).

The improvements are:
* Recycling of `these_discoveries` object
* ~Bulk update of discoveries~ (removed after the last improvement)
* Bulk insert of discoveries
* Insert of discoveries at the end of the scan

## Initial performance assessment

<details>
<summary>Details</summary>

```bash
Clone repo - 4.4131 seconds.

Scan commits - 15.6877 seconds, mean: 0.0391 seconds, variance: 0.0005 seconds
        Diff hash - mean: 0.0001 seconds, variance: 0.0 seconds.
        Diff - mean: 0.0372 seconds, variance: 0.0003 seconds.
        Worker - mean: 0.0018 seconds, variance: 0.0001 seconds.

Insert discoveries - 15.6187 seconds.

ML models - 14.9182 seconds.
        PathModel - 7.6944 seconds. mean: 0.0011 seconds, variance: 0.0 seconds.
        SnippetModel - 7.2237 seconds. mean: 0.0009 seconds, variance: 0.0 seconds.

Total scan time: 53.1926 seconds.
```

</details>

Tests have been made on a fork of this repository, so that code would remain the same in every test.
The repo has 8 branches, ~2k commits and ~6k discoveries.

Initial total scan time was ~53 seconds, and it was comprised of:
* Repo clone: ~4s
* Commits scan: ~15s
* Insert discoveries: ~15s
* ML models application: ~15s

### Some considerations
* Nothing can be done about cloning the repository at this stage.
* I tried applying multithread and multiprocessing to the commits scan but the project's dependencies do not support it, so nothing can be done about it at this stage.
* Insert discoveries takes a long time and can be improved.
* ML models application takes a long time, and it is IO bound at this stage.

Also, the time performance scales with two factors at this stage:
* number of commits: the scan has to go through each commit and analyze it. This is why I tried to apply multiprocessing. (this is not included in this PR)
* number of discoveries: queries to the database are not optimized, so a query per discovery is made. This PR targets this scale factor.

## Recycling of `these_discoveries` object

<details>
<summary>Details</summary>

```bash
Clone repo - 2.5628 seconds.

Scan commits - 16.7548 seconds, mean: 0.0418 seconds, variance: 0.0006 seconds
        Diff hash - mean: 0.0001 seconds, variance: 0.0 seconds.
        Diff - mean: 0.0396 seconds, variance: 0.0004 seconds.
        Worker - mean: 0.002 seconds, variance: 0.0001 seconds.

Insert discoveries - 14.123 seconds.

ML models - 4.5517 seconds.
        PathModel - 2.8357 seconds. mean: 0.0004 seconds, variance: 0.0 seconds.
        SnippetModel - 1.716 seconds. mean: 0.0003 seconds, variance: 0.0 seconds.

Total scan time: 40.6276 seconds.
```

</details>

First I added bulk update to the discoveries by recycling the (already in memory) `these_discoveries` object, bringing down ML models time by ~10s.
Recycling means that, once the ML models have been applied to the discoveries, I am updating the in-memory object with the new state, and then I am updating it in the db once all discoveries have passed the ML models, doing a bulk update and reducing the queries to the db from ~6k (one per discovery) to ~1.

## Bulk insert of discoveries

<details>
<summary>Details</summary>

```bash
Clone repo - 2.7995 seconds.

Scan commits - 16.0279 seconds, mean: 0.04 seconds, variance: 0.0005 seconds
        Diff hash - mean: 0.0001 seconds, variance: 0.0 seconds.
        Diff - mean: 0.0378 seconds, variance: 0.0003 seconds.
        Worker - mean: 0.0021 seconds, variance: 0.0001 seconds.

Insert discoveries - 0.2585 seconds.

ML models - 4.5205 seconds.
        PathModel - 2.8357 seconds. mean: 0.0004 seconds, variance: 0.0 seconds.
        SnippetModel - 1.6848 seconds. mean: 0.0002 seconds, variance: 0.0 seconds.

Total scan time: 25.823 seconds.
```

</details>

Then I added bulk insert of discoveries, bringing down insert discoveries time by ~100% and reducing queries to the db from ~6k (one per discovery) to ~1.

## Move discoveries insert at the end of the scan

<details>
<summary>Details</summary>

```bash
Clone repo - 3.4692 seconds.

Scan commits - 15.2663 seconds, mean: 0.0256 seconds, variance: 0.0003 seconds
        Diff hash - mean: 0.0 seconds, variance: 0.0 seconds.
        Diff - mean: 0.0235 seconds, variance: 0.0001 seconds.
        Worker - mean: 0.0021 seconds, variance: 0.0001 seconds.

ML models - 4.4551 seconds.
        PathModel - 2.9736 seconds. mean: 0.0005 seconds, variance: 0.0 seconds.
        SnippetModel - 1.4815 seconds. mean: 0.0002 seconds, variance: 0.0 seconds.

Insert discoveries - 0.3299 seconds.

Total scan time: 25.1816 seconds.
```

</details>

Finally, I moved the discoveries db insertion at the end of the scan, so that we wouldn't need to first insert them and then update them when the ML models would update their state. With this, the change of state happens in memory, so that when the discoveries get inserted into the db, they already present the correct state.
Unfortunately, this seems to have no change in performance, due to the fact that I already had optimized the update with bulk update.

### CODE REVIEW

**Comment by fabiosangregorio** [comment]
@marcorosa I fixed the error and the PR is now ready for review.

File: credentialdigger/client.py [hunk 1]
```diff
@@ -447,6 +451,27 @@ def update_discovery(self, query, discovery_id, new_state):
...
 
+    def update_discoveries(self, query, discoveries_ids, new_state):
+        """ Change the state of multiple discoveries.
+
+        Parameters
+        ----------
+        discoveries_ids: list
+            The ids of the discoveries to be updated
+        new_state: str
+            The new state of these discoveries
```
Comment (marcorosa) [inline credentialdigger/client.py:462]: here you miss the parameter `query`

**Review by marcorosa — CHANGES_REQUESTED** [review]

File: credentialdigger/client.py (hunk 1 above)
Comment (fabiosangregorio) [reply credentialdigger/client.py:462]: @marcorosa parameter `query` is missing in all methods of `client.py`.  Shall I add it to each method?

Comment (marcorosa) [reply credentialdigger/client.py:462]: yes, please

**Review by marcorosa — APPROVED** [review]

**Pull request merged**